feat: 'Try this' support

digama0 opened the pull request:
Adds proper widget / code action support for "try this" prompts in tactics. Still open: merging the widget into the message itself, depends on leanprover/lean4#2064 .

fpvandoorn commented on Std/Tactic/TryThis.lean:
```diff
@@ -1,42 +1,126 @@
+  }
+  return e('div', {className: 'ml1'}, e('pre', {className: 'font-code pre-wrap'}, [
+    'Try this: ',
+    e('a', {onClick, className: 'link pointer dim', title: 'Apply suggestion'}, props.suggestion),
```
Could you add `style: { cursor: 'pointer' }` to the `onClick` structure, so that the mouse becomes a pointer on hover? (suggested by Adam Topaz in the Zulip thread)

fpvandoorn commented:
Two more things I noticed: 
* The `tryThisAt` in your original code snippet doesn't produce a blue squiggle under the tactic. Is it possible to add this (currently I also use a `logInfoAt` to produce the squiggle)
* In the implementation of `recommend using h` I wrote, I put the `logInfoAt` and `tryThisAt` both on the word `recommend`. If I have my pointer at the end of the line, I still see the `logInfoAt` message that was created on the same line (without opening "All messages"). However, to see the tactic replacements, I need to have my pointer exactly on the word `recommend`. Ideally it would show it whenever you are on the same line.
![image](https://user-images.githubusercontent.com/7376012/237047468-2317cdfd-9bfb-4c68-bb53-d4a8a0ba7952.png)

gebner commented on Std/Tactic/TryThis.lean:
```diff
@@ -1,42 +1,126 @@
+def addSuggestion (origStx : Syntax) {kind : Name} (suggestion : TSyntax kind)
+    (suggestionForMessage : Option MessageData := none)
+    (ref? : Option Syntax := none)
+    (extraMsg : String := "") : MetaM Unit := do
```
I think it would be clearer to swap (the names) `origStx` and `ref?` here.  Usually we use `ref` to specify where the error message should go.  While "original syntax" sounds like it should be the original tactic/term that we want to replace using the suggestion.

digama0 replied:
I reorganized the names a bit and added parameter documentation.

digama0 commented:
@fpvandoorn See whether the new version addresses your concerns. The widget hover (and code action) now happens as long as your cursor is over any of the lines containing the message, same as the messages display.

Vtec234 commented on Std/Tactic/TryThis.lean:
```diff
@@ -1,42 +1,144 @@
+/-- Add a `exact e` or `refine e` suggestion.
+
+The parameters are:
+* `ref`: the original syntax to be replaced
```
```suggestion
* `ref`: the span of the info diagnostic
```
Right?

Vtec234 commented on Std/Tactic/TryThis.lean:
```diff
@@ -1,42 +1,144 @@
+/-- Add a term suggestion.
+
+The parameters are:
+* `ref`: the original syntax to be replaced
```
```suggestion
* `ref`: the span of the info diagnostic
```

Vtec234 commented on Std/Tactic/TryThis.lean:
```diff
@@ -1,42 +1,144 @@
+* `ref`: the span of the info diagnostic
+* `suggestion`: the replacement syntax
+* `suggestionForMessage?`: the message to display in the info diagnostic (only).
+  The widget message uses only `suggestion`. If not provided, it uses `suggestion` in both places.
```
```suggestion
  The widget message uses only `suggestion`. If not provided, `suggestion` is used in both places.
```
Otherwise it sounds like "the widget message uses `suggestion` in both places" which is presumably not what was meant.

Vtec234 reviewed (approved):
Looks good, all functional issues with the PR that we identified are to be fixed in core rather than here.
